src/Syntax/Types.hs

```diff
@@ -0,0 +1,161 @@
+
+import Data.Singletons
+import Data.Type.Equality (TestEquality(..))
+import Data.Typeable hiding (TypeRep,typeRep)
+import Type.Reflection
+
+import Data.ByteString    as Syntax.Types (ByteString)
+import EVM.ABI            as Syntax.Types (AbiType(..))
+
+-- | Types understood by proving tools.
+data MType
+  = Integer
+  | Boolean
+  | ByteStr
+  deriving (Eq, Ord, Show, Read)
+
+-- | Singleton types of the types understood by proving tools.
+data SType a where
+  SInteger :: SType Integer
+  SBoolean :: SType Bool
+  SByteStr :: SType ByteString
+deriving instance Eq (SType a)
+
+instance Show (SType a) where
+  show = \case
+    SInteger -> "int"
+    SBoolean -> "bool"
+    SByteStr -> "bytestring"
+
+instance TestEquality SType where
+  testEquality STypeable STypeable = eqT
+
+eqS :: forall (a :: *) (b :: *) f t. (SingI a, SingI b, Eq (f a t)) => f a t -> f b t -> Bool
+eqS fa fb = maybe False (\Refl -> fa == fb) $ testEquality (sing @a) (sing @b)
+
+class HasType a t where
+  getType :: a -> SType t
+
+metaType :: AbiType -> MType
+metaType (AbiUIntType _)     = Integer
+metaType (AbiIntType  _)     = Integer
+metaType AbiAddressType      = Integer
+metaType AbiBoolType         = Boolean
+metaType (AbiBytesType n)    = if n <= 32 then Integer else ByteStr
+metaType AbiBytesDynamicType = ByteStr
+metaType AbiStringType       = ByteStr
+--metaType (AbiArrayDynamicType a) =
+--metaType (AbiArrayType        Int AbiType
+--metaType (AbiTupleType        (Vector AbiType)
+metaType _ = error "Syntax.Types.metaType: TODO"
+
+-- | For our purposes, the singleton of a type @a@ is always @'SType' a@.
+-- Note that even though there only exist three different 'SType', this does
+-- not mean that the type family is partial. It simply means that the resulting
+-- type is uninhabited if the argument is neither 'Integer', 'Bool' nor
+-- 'ByteString'.
+type instance Sing = SType
+
+-- Defines which singleton to retreive when we only have the type, not the
+-- actual singleton.
+instance SingI Integer    where sing = SInteger
+instance SingI Bool       where sing = SBoolean
+instance SingI ByteString where sing = SByteStr
+
+-- | This instance allows us to go between 'MType', @'SType' a@ and @a@,
+-- with @a :: '*'@.
+instance SingKind * where
+  -- | We can demote a type variable @a@ to a value of type 'MType'
+  type Demote * = MType
+
+  -- | We can go from any singleton type to the corresponding demoted type.
+  fromSing SInteger = Integer
+  fromSing SBoolean = Boolean
+  fromSing SByteStr = ByteStr
+
+  -- | We can go from any demoted type to the corresponding singleton type,
+  -- but need to hide its type variable when doing so.
+  toSing Integer = SomeSing SInteger
+  toSing Boolean = SomeSing SBoolean
+  toSing ByteStr = SomeSing SByteStr
+
+-- | Pattern match on an 'EVM.ABI.AbiType' is if it were an 'SType' with a 'Typeable'
+-- instance.
+pattern FromAbi :: () => Typeable a => SType a -> AbiType
```

Comment:
We are defining a custom pattern `FromAbi` that we can match against. This will create something that behaves like a constructor at pattern matching (but we can't use it to *construct* values, only deconstruct).

Specifically, if we have a value of type `AbiType`, this allows us to get the corresponding `Typeable a => SType a`.

The way this works is in three steps: 
1. Use `metaType :: AbiType -> MType`
2. Use `toSing :: MType -> SType a` from lines 96-98. But instead of using it directly, we use the convenience pattern `FromSing` from the `singletons` library.
3. Match the result on our own `STypeable` pattern with an `@`-pattern. This retrieves the `Typeable a` instance for any `SType a`, which is possible since we know all the cases of `SType`.